pinch-zoom-canvas.js

```diff
@@ -1,468 +1,528 @@
+        if (this.imgTexture.width) {
+          console.log('canvas', this.canvas.width, this.canvas.height)
+
+          var viewportRatio = this.canvas.width / this.canvas.height
+          var imageRatio = this.imgTexture.width / this.imgTexture.height
+          var scaleRatio = null
+
+          if (imageRatio >= viewportRatio) { // wide image
+            this.initResizeProperty = 'width'
+            scaleRatio = this.canvas.width / this.imgTexture.width * this.startingZoom // startingZoom multiplier
+          } else if (imageRatio < viewportRatio) { // tall image
+            this.initResizeProperty = 'height'
+            scaleRatio = this.canvas.height / this.imgTexture.height * this.startingZoom // startingZoom multiplier
+          }
+
+          this.position.x = (this.canvas.width - this.imgTexture.width * scaleRatio) / 2 // center horizontal
+          this.position.y = (this.canvas.height - this.imgTexture.height * scaleRatio) / 2 // center vertical
+
+          // scale x and y to init scaleRatio calculation
+          this.scale.x = scaleRatio
+          this.scale.y = scaleRatio
+
+          // initial position is centered in the canvas
+          this.initPosition = {
+            x: this.position.x,
+            y: this.position.y
+          }
+
+          // the initial scale is the scaling ratio
+          this.initialScale = scaleRatio // includes the startingZoom
+          this.calculateOffset()
+
+          // start the impetus so we can move things right away if using momentum
+          if (this.momentum) this._createImpetus()
+
+          this.init = true // done initializing!
         }
+      }
+
+      // erases the canvas
+      this.context.clearRect(0, 0, this.canvas.width, this.canvas.height)
+
+      // draws the image position and scale
+      this.context.drawImage(
+        this.imgTexture,
+        this.position.x, this.position.y,
+        this.scale.x * this.imgTexture.width,
+        this.scale.y * this.imgTexture.height)
+
+      requestAnimationFrame(this.render)
+    },
+
+    pause: function () {
+      this.running = false
+      return this
+    },
+
+    resume: function () {
+      this.calculateOffset()
+      this.running = true
+      requestAnimationFrame(this.render)
+      return this
+    },
+
+		/**
+		 * Calculates the offset of the canvas position relative to the page
+     * since touch events are always relative to the page
+		 */
+    calculateOffset: function () {
+      if (!this.canvas) return this
+
+      // no offset if we're using direct client coordinates (full screen canvas)
+      if (!this.fullScreen) {
+        const canvasBox = this.canvas.getBoundingClientRect()
+
+        // caculate the offset from the scroll position
+        this.offset.x = canvasBox.left + window.scrollX
+        this.offset.y = canvasBox.top + window.scrollY
+      }
+
+      return this
+    },
+
+    /**
+     * handles zooming in and out
+     */
+    zoom: function (zoom, touchX, touchY) {
```

Comment:
this should remain public as you can call it externally to force move things outside of user input